Add HTTP endpoint support

lekv opened the pull request:
This is a preview of adding HTTP endpoint support (#357). I tested this locally against an Impala cluster on my development machine and added a trivial test to it. Additional ideas on how to test this would be very much welcome.

lekv commented:
@timarmstrong Do you have time to take a look at this one?

timarmstrong commented on impala/_thrift_api.py:
```diff
@@ -120,6 +123,43 @@ def get_socket(host, port, use_ssl, ca_cert):
+    if use_ssl:
+        url = 'https://%s:%s/%s' % (host, port, http_path)
+        log.debug('get_http_transport url=%s', url)
+        # TODO: Add server authentication with thrift 0.12
```
Should probably raise an exception if ca_cert was not provided and we're not checking it

lekv replied:
Done, I added the check to the calling code because I didn't want to introduce dbapi errors to the thrift api wrapper.

timarmstrong commented on impala/hiveserver2.py:
```diff
@@ -758,39 +759,47 @@ def threaded(func):
-    if krb_host:
-        kerberos_host = krb_host
+    if use_http_transport:
+        transport = get_http_transport(host, port, http_path=http_path,
```
Do you need to pass in auth_mechanism?

lekv replied:
Uh yes, fixed

timarmstrong commented on impala/hiveserver2.py:
```diff
@@ -758,39 +759,47 @@ def threaded(func):
-    if krb_host:
-        kerberos_host = krb_host
+    if use_http_transport:
+        transport = get_http_transport(host, port, http_path=http_path,
```
Validate that we don't have kerberos enabled for http?

lekv replied:
Done

lekv commented:
Addressed the comments and tested the client using this change of Impala: https://gerrit.cloudera.org/#/c/14059/

I briefly looked into the LDAP test code but haven't had time to add a proper test.

timarmstrong commented on impala/_thrift_api.py:
```diff
@@ -120,6 +123,43 @@ def get_socket(host, port, use_ssl, ca_cert):
+            else:
+                # PLAIN always requires a password for HS2.
+                password = 'password'
+            log.debug('get_http_transport: password=%s', password)
```
It doesn't look like we use password, or am I missing something - it's all ldap_password below.

lekv replied:
Fixed this one

timarmstrong commented on impala/hiveserver2.py:
```diff
@@ -758,39 +759,55 @@ def threaded(func):
-    if krb_host:
-        kerberos_host = krb_host
+    if use_http_transport:
+        # TODO: Add server authentication with thrift 0.12
```
Is there a thrift JIRA or something we can link to?

lekv replied:
Opened #362 and added it to the comment.

lekv commented:
Thanks for the comments. I added some LDAP tests to the Impala change that I used to test this: https://gerrit.cloudera.org/#/c/14059/

Once this is in, we will have to release a new version of Impyla, then we can bump the dependency in Impala and get that change in, too.

dknupp commented on impala/hiveserver2.py:
```diff
@@ -758,29 +759,46 @@ def threaded(func):
 def connect(host, port, timeout=None, use_ssl=False, ca_cert=None,
             user=None, password=None, kerberos_service_name='impala',
-            auth_mechanism=None, krb_host=None):
+            auth_mechanism=None, krb_host=None, use_http_transport=False,
```
@lekv -- What is the difference between auth_mechanism=None and auth_mechanism=NOSASL? dbapi.py specifies these values:

`AUTH_MECHANISMS = ['NOSASL', 'PLAIN', 'GSSAPI', 'LDAP']`

NOSASL in the default in the connect() method from that file.
